File: completed/1233.py
class Solution:
    def removeSubfolders(self, folder):
        parents = set()
        for path in sorted(folder):
            exists = False
            components = path.split("/")
            for i in range(1, len(components)):
                if "/".join(components[:i]) in parents:
                    exists = True
                    break
            if not exists:
                parents.add(path)

        return list(parents)
        

if __name__ == "__main__":
    # folder = [ "/a", "/a/b", "/c/d", "/c/d/e", "/c/f" ]
    folder = [ "/a/b/c", "/a/b/ca", "/a/b/d" ]

    solution = Solution()
    print(solution.removeSubfolders(folder))

File: completed/1233.java
import java.util.*;

class Solution {
    private class Node {
        boolean isTerminal;
        Map<String, Node> nextNodes;

        Node() {
            this.nextNodes = new HashMap<>();
        }

        boolean add(String[] components, int index) {
            if (index == components.length) {
                return true;
            }

            String component = components[index];
            Node nextNode = nextNodes.get(component);
            if (nextNode == null) {
                nextNode = new Node();
                nextNode.isTerminal = (index == components.length - 1);
                nextNodes.put(component, nextNode);
            } else if (nextNode.isTerminal) {
                return false;
            }
            return nextNode.add(components, index + 1);
        }
    }

    public List<String> removeSubfolders(String[] folder) {
        Node root = new Node();
        List<String> results = new ArrayList<>();
        Arrays.sort(folder);
        for (String path : folder) {
            String[] components = path.split("/");
            if (root.add(components, 0)) {
                results.add(path);
            }
        }
        return results;
    }
}
